contracts/registry/src/lib.rs

```diff
@@ -86,29 +86,105 @@ impl Contract {
                     self.banlist.insert(&owner),
                     "caller banned: can't make soul transfer"
                 );
-                require!(!self._is_banned(&to), "`to` is banned");
-                emit_soul_transfer(&owner, &to);
-                IssuerTokenId {
-                    issuer_id: 0,
-                    token: 0,
-                }
+                (
+                    false,
+                    IssuerTokenId {
+                        issuer_id: 0,
+                        token: 0, // NOTE: this is class ID
+                    },
+                )
             }
             // resuming Soul Transfer process
-            Some(s) => s,
+            Some(s) => (true, s),
         };
 
-        println!("Starting at: {} {}", start.issuer_id, start.token);
-        env::panic_str("not implemented");
+        let batch: Vec<(BalanceKey, TokenId)> = self
+            .balances
+            .iter_from(BalanceKey {
+                owner: owner.clone(),
+                issuer_id: start.issuer_id,
+                class_id: start.token,
+            })
+            .take(limit)
+            .collect();
+
+        let mut b_new = BalanceKey {
+            owner: to.clone(),
+            issuer_id: 0,
+            class_id: 0,
+        };
+        let mut prev_issuer: IssuerId = 0;
+        let mut i = 0;
+        for (b, tid) in &batch {
+            if b.owner != owner {
+                break;
+            }
+            i += 1;
+
+            if prev_issuer != b.issuer_id {
+                prev_issuer = b.issuer_id;
+                // update user token supply map
+                match self.supply_by_owner.remove(&(owner.clone(), prev_issuer)) {
+                    None => (),
+                    Some(supply_owner) => {
+                        let key = &(to.clone(), prev_issuer);
+                        let supply_to = self.supply_by_owner.get(key).unwrap_or(0);
+                        self.supply_by_owner
+                            .insert(key, &(supply_owner + supply_to));
+                    }
+                }
+            }
+
+            self.balances.remove(b);
+            b_new.issuer_id = b.issuer_id;
+            b_new.class_id = b.class_id;
+            // TODO: decide if we should overwrite or panic if receipient already had a token.
+            // now we overwrite.
+            self.balances.insert(&b_new, tid);
+            self.balances.remove(&b);
+
+            let i_key = IssuerTokenId {
+                issuer_id: b.issuer_id,
+                token: tid.clone(),
+            };
+            let mut td = self.issuer_tokens.get(&i_key).unwrap();
+            td.owner = to.clone();
+            self.issuer_tokens.insert(&i_key, &td);
+        }
+
+        let completed = i != limit;
+        // we emit the event only once the operation is completed
+        if completed {
+            self.ongoing_soul_tx.remove(&owner);
+            if resumed || i > 0 {
```

Comment:
Do we also emit the event if the transfer is already ongoing? Also shouldn't we also insert the current transfer we are doing here to the map as well? We are removing it here but I dont see it being inserted. I might have missed something

Reply:
The event should be emitted only once.

>  Also shouldn't we also insert the current transfer we are doing here to the map as well?

You mean the `ongoing_soul_tx`? Yes, I forgot about it.